distributed/src/main/scala/asobu/distributed/service/Extractors.scala

```diff
@@ -0,0 +1,196 @@
+package asobu.distributed.service
+
+import asobu.distributed.service.Action.DistributedRequest
+import asobu.distributed.service.Extractors._
+import asobu.distributed.RequestExtractorDefinition
+import asobu.distributed.gateway.SyncedExtractResult
+import asobu.distributed.service.Action.DistributedRequest
+import asobu.distributed.service.Extractors.{RouteParamsExtractor, BodyExtractor, RemoteExtractor}
+import asobu.dsl._
+import asobu.dsl.extractors.JsonBodyExtractor
+import asobu.dsl.util.HListOps.{RestOf, CombineTo, RestOf2}
+import cats.{Monad, Functor, Eval}
+import cats.sequence.RecordSequencer
+import shapeless.ops.hlist.Prepend
+import asobu.dsl.util.RecordOps.{FieldKV, FieldKVs}
+import cats.data.{Xor, Kleisli}
+import play.api.libs.json.{Reads, Json}
+import play.core.routing.RouteParams
+import shapeless.labelled.FieldType
+import shapeless.ops.hlist.Mapper
+import shapeless._, labelled.field
+import ExtractResult._
+import cats.syntax.all._
+import CatsInstances._
+import play.api.mvc._, play.api.mvc.Results._
+
+import scala.annotation.implicitNotFound
+import scala.concurrent.Future
+
+trait Extractors[TMessage] {
+  /**
+   * List to be extracted from request and send to the remote handler
+   */
+  type LToSend <: HList
+  type LParam <: HList
+  type LExtra <: HList
+
+  val remoteExtractorDef: RemoteExtractorDef[LToSend, LParam, LExtra]
+
+  def localExtract(dr: DistributedRequest[LToSend]): ExtractResult[TMessage]
+}
+
+object Extractors {
+  type Aux[TMessage, LToSend0 <: HList, LParam0 <: HList, LExtra0 <: HList] = Extractors[TMessage] {
+    type LToSend = LToSend0
+    type LParam = LParam0
+    type LExtra = LExtra0
+  }
+
+  type RouteParamsExtractor[T] = Kleisli[SyncedExtractResult, RouteParams, T]
+
+  /**
+   * Extract information at the gateway end
+   */
+  type RemoteExtractor[T] = Extractor[(RouteParams, Request[AnyContent]), T]
+
+  type BodyExtractor[T] = Extractor[AnyContent, T]
+
+  class builder[TMessage] {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Class name does not match the regular expression '^[A-Z][A-Za-z0-9]*$'](https://www.codacy.com/app/iHeartRadio/asobu/file/2312372375/issues/source?bid=3037259&fileBranchId=3178877#l59)